cmd/snap-bootstrap/cmd_initramfs_mounts.go

```diff
@@ -908,6 +916,16 @@ func (m *recoverModeStateMachine) unlockSaveFallbackKey() (stateFunc, error) {
 	// remember what we assumed about encryption before looking at
 	// save
 	assumeEncrypted := m.isEncryptedDev
+	// do we even have ubuntu-data?
+	haveData := m.degradedState.partition("ubuntu-data").FindState == partitionFound
+
+	if m.noFallback && (!haveData || assumeEncrypted) {
```

Comment:
this logic is a bit confusing to me what is the purpose of the `|| assumeEncrypted)` ?

Also I think this may warrant at least an explanation in the README about the behavior if not an update to the state machine diagram (though I don't think there are any new transitions, just an additional exit path)

Reply:
I don't understand why we have  a && part here at all? which scenario are we considering? I'm not understanding the comment below

Reply:
Those are 2 possible scenarios where we want to exit early:

- the system is encrypted (`assumeEncrypted == true`), the call to using fallback key is unwarranted for whatever reason (unlocking the data or save with run key must have failed, or data is not mounted)
- we could not find ubuntu-data (`!haveData == true`), thus we don't know if the system is encrypted or not and we try using save fallback key, but we don't want that either

I have left out the case for system being unencrypted and data failing to mount, as I assume this is part of the health check, which basically tells whether the relevant data is there (in this case it's state.json). Though to make it consistent in all cases, the check could verify that data is actually mounted and block this path early too. Does that make sense?

Reply:
As discussed in a quick chat, I'll try to prepare a little cleanup of the state functions around this part so that it is clear whether we are following the encrypted or unencrypted device code paths.